---
title: "Seurat Output"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
params:
  project: ProjectTitle
  path:
    label: "Copy and paste 10X gene barcode file path"
    value: "data/outs/filtered_gene_bc_matrices/hg19"
    input: text
  cells:
    label: "Include genes with this many cells"
    value: 3
    input: slider
    min: 1
    max: 20
  genes:
    label: "Include cells where at least this many genes are detected"
    value: 200
    input: slider
    min: 10
    max: 500
  max_mt:
    label: "Max mitochondrial genes present"
    value: 0.05
    input: numeric
    min: 0
    max: 1
    step: 0.01
---
## Introduction

This is literally a copy and paste of this [Seurat tutorial](http://satijalab.org/seurat/pbmc3k_tutorial.html) meant as a first pass at 10x data.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup the Seurat Object

The original tutorial analyzed a dataset of Peripheral Blood Mononuclear Cells (PBMC) freely available from 10X Genomics. There are 2,700 single cells that were sequenced on the Illumina NextSeq 500. The raw data can be found [here](https://s3-us-west-2.amazonaws.com/10x.files/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).

We start by reading in the data. All features in Seurat have been configured to work with sparse matrices which results in significant memory and speed savings for Drop-seq/inDrop/10x data.

```{r SeuratObject, echo = FALSE, message = FALSE, warning = FALSE}
library(Seurat)
library(dplyr)
library(Matrix)

# Load the dataset
raw_data <- Read10X(data.dir = params$path)

# Initialize the Seurat object with the raw (non-normalized data).  Keep all
# genes expressed in >= 3 cells (~0.1% of the data). Keep all cells with at
# least 200 detected genes
obj <- CreateSeuratObject(raw.data = raw_data, 
                          min.cells = params$cells, 
                          min.genes = params$genes, 
                          project = params$project)
```

## Standard pre-processing workflow

The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the creation of a Seurat object, the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable genes. In previous versions, we grouped many of these steps together in the <code>Setup</code> function, but in v2, we separate these steps into a clear and sequential workflow.

## QC and selecting cells for further analysis

While the <code>CreateSeuratObject</code> imposes a basic minimum gene-cutoff, you may want to filter out cells at this stage based on technical or biological parameters. Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. In the example below, we visualize gene and molecule counts, plot their relationship, and exclude cells with a clear outlier number of genes detected as potential multiplets. Of course this is not a guaranteed method to exclude cell doublets, but we include this as an example of filtering user-defined outlier cells. We also filter cells based on the percentage of mitochondrial genes present.


```{r QCSelection}
# The number of genes and UMIs (nGene and nUMI) are automatically calculated
# for every object by Seurat.  For non-UMI data, nUMI represents the sum of
# the non-normalized values within a cell We calculate the percentage of
# mitochondrial genes here and store it in percent.mito using AddMetaData.
# We use object@raw.data since this represents non-transformed and
# non-log-normalized counts The % of UMI mapping to MT-genes is a common
# scRNA-seq QC metric.  NOTE: You must have the Matrix package loaded to
# calculate the percent.mito values.
mito.genes <- grep(pattern = "^MT-", x = rownames(x = obj@data), value = TRUE)
percent.mito <- Matrix::colSums(obj@raw.data[mito.genes, ])/Matrix::colSums(obj@raw.data)

# AddMetaData adds columns to object@meta.data, and is a great place to
# stash QC stats
obj <- AddMetaData(object = obj, metadata = percent.mito, col.name = "percent.mito")
VlnPlot(object = obj, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)
```

GenePlot is typically used to visualize gene-gene relationships, but can
be used for anything calculated by the object, i.e. columns in
object@meta.data, PC scores etc.  Since there is a rare subset of cells
with an outlier level of high mitochondrial percentage and also low UMI
content, we filter these as well

```{r GenePlot}
par(mfrow = c(1, 2))
GenePlot(object = obj, gene1 = "nUMI", gene2 = "percent.mito")
GenePlot(object = obj, gene1 = "nUMI", gene2 = "nGene")
```

We filter out, by default, cells that have unique gene counts over 2,500 or less than
200 Note that low.thresholds and high.thresholds are used to define a
'gate' -Inf and Inf should be used if you don't want a lower or upper
threshold.  These can be changed from default when first rendering the document.

```{r FilterThreshold}

obj <- FilterCells(object = obj, subset.names = c("nGene", "percent.mito"), 
    low.thresholds = c(params$genes, -Inf), high.thresholds = c(Inf, params$max_mt))
```


## Normalizing the data
After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.

```{r NormalizeData}
obj <- NormalizeData(object = obj, 
                      normalization.method = "LogNormalize", 
                      scale.factor = 10000)
```

## Detection of variable genes across the single cells

Seurat calculates highly variable genes and focuses on these for downstream analysis. FindVariableGenes calculates the average expression and dispersion for each gene, places these genes into bins, and then calculates a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko et al.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~2,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.


```{r DetectVarGenes}
obj <- FindVariableGenes(object = obj, 
                         mean.function = ExpMean, 
                         dispersion.function = LogVMR, 
                         x.low.cutoff = 0.0125, 
                         x.high.cutoff = 3, 
                         y.cutoff = 0.5)

length(x = obj@var.genes)
```


## Scaling the data and removing unwanted sources of variation

Your single cell dataset likely contains ‘uninteresting’ sources of variation. This could include not only technical noise, but batch effects, or even biological sources of variation (cell cycle stage). As suggested in Buettner et al, NBT, 2015, regressing these signals out of the analysis can improve downstream dimensionality reduction and clustering. To mitigate the effect of these signals, Seurat constructs linear models to predict gene expression based on user-defined variables. The scaled z-scored residuals of these models are stored in the scale.data slot, and are used for dimensionality reduction and clustering.

We can regress out cell-cell variation in gene expression driven by batch (if applicable), cell alignment rate (as provided by Drop-seq tools for Drop-seq data), the number of detected molecules, and mitochondrial gene expression. For cycling cells, we can also learn a ‘cell-cycle’ score (see example [HERE]) and regress this out as well. In this simple example here for post-mitotic blood cells, we regress on the number of detected molecules per cell as well as the percentage mitochondrial gene content.

Seurat v2.0 implements this regression as part of the data scaling process. Therefore, the <code>RegressOut</code> function has been deprecated, and replaced with the vars.to.regress argument in <code>ScaleData</code>.

```{r ScaleData}
obj <- ScaleData(object = obj, vars.to.regress = c("nUMI", "percent.mito"))
```


## Perform linear dimensional reduction

Next we perform PCA on the scaled data. By default, the genes in <code>object@var.genes</code> are used as input, but can be defined using pc.genes. We have typically found that running dimensionality reduction on highly variable genes can improve performance. However, with UMI data - particularly after regressing out technical variables, we often see that PCA returns similar (albeit slower) results when run on much larger subsets of genes, including the whole transcriptome.

Seurat provides several useful ways of visualizing both cells and genes that define the PCA, including <code>PrintPCA</code>, <code>VizPCA</code>,  <code>PCAPlot</code>, and <code>PCHeatmap</code>

### PrintPCA

```{r RunPCA}
obj <- RunPCA(object = obj, 
              pc.genes = obj@var.genes, 
              do.print = TRUE, 
              pcs.print = 1:5, 
              genes.print = 5)
```

### VizPCA

```{r VizPCA}
VizPCA(object = obj, pcs.use = 1:2)
```

### PCA Plot

```{r PCAPlot}
PCAPlot(object = obj, dim.1 = 1, dim.2 = 2)
```

### Project PCA

ProjectPCA scores each gene in the dataset (including genes not included
in the PCA) based on their correlation with the calculated components.
Though we don't use this further here, it can be used to identify markers
that are strongly correlated with cellular heterogeneity, but may not have
passed through variable gene selection.  The results of the projected PCA
can be explored by setting use.full=T in the functions above

```{r ProjectPCA}
obj <- ProjectPCA(object = obj, do.print = FALSE)
```

### PCA Heatmap

In particular PCHeatmap allows for easy exploration of the primary sources of heterogeneity in a dataset, and can be useful when trying to decide which PCs to include for further downstream analyses. Both cells and genes are ordered according to their PCA scores. Setting cells.use to a number plots the ‘extreme’ cells on both ends of the spectrum, which dramatically speeds plotting for large datasets. Though clearly a supervised analysis, we find this to be a valuable tool for exploring correlated gene sets.

```{r PCAHeatmap}
PCHeatmap(object = obj, pc.use = 1, cells.use = 500, do.balanced = TRUE, label.columns = FALSE)
```

### Heatmaps by PCs

```{r PCHeatmaps}
PCHeatmap(object = obj, 
          pc.use = 1:12, 
          cells.use = 500, 
          do.balanced = TRUE, 
          label.columns = FALSE, 
          use.full = FALSE)
```

## Determine statistically significant principal components
To overcome the extensive technical noise in any single gene for scRNA-seq data, Seurat clusters cells based on their PCA scores, with each PC essentially representing a ‘metagene’ that combines information across a correlated gene set. Determining how many PCs to include downstream is therefore an important step.

In Macosko et al, we implemented a resampling test inspired by the jackStraw procedure. We randomly permute a subset of the data (1% by default) and rerun PCA, constructing a ‘null distribution’ of gene scores, and repeat this procedure. We identify ‘significant’ PCs as those who have a strong enrichment of low p-value genes.

```{r JackStraw}
# NOTE: This process can take a long time for big datasets, comment out for
# expediency.  More approximate techniques such as those implemented in
# PCElbowPlot() can be used to reduce computation time
obj <- JackStraw(object = obj, num.replicate = 100, do.print = FALSE)
```


The JackStrawPlot function provides a visualization tool for comparing the distribution of p-values for each PC with a uniform distribution (dashed line). ‘Significant’ PCs will show a strong enrichment of genes with low p-values (solid curve above the dashed line). In this case it appears that PCs 1-10 are significant.

```{r JackStrawPlot}
JackStrawPlot(object = obj, PCs = 1:12)
```


A more ad hoc method for determining which PCs to use is to look at a plot of the standard deviations of the principle components and draw your cutoff where there is a clear elbow in the graph. This can be done with PCElbowPlot. In this example, it looks like the elbow would fall around PC 9.

```{r PCEPlot}
PCElbowPlot(object = obj)
```

PC selection – identifying the true dimensionality of a dataset – is an important step for Seurat, but can be challenging/uncertain for the user. We therefore suggest these three approaches to consider. The first is more supervised, exploring PCs to determine relevant sources of heterogeneity, and could be used in conjunction with GSEA for example. The second implements a statistical test based on a random null model, but is time-consuming for large datasets, and may not return a clear PC cutoff. The third is a heuristic that is commonly used, and can be calculated instantly. In this example, all three approaches yielded similar results, but we might have been justified in choosing anything between PC 7-10 as a cutoff. We followed the jackStraw here, admittedly buoyed by seeing the PCHeatmap returning interpretable signals (including canonical dendritic cell markers) throughout these PCs. Though the results are only subtly affected by small shifts in this cutoff (you can test below), we strongly suggest always explore the PCs they choose to include downstream.

## Cluster the cells

Seurat now includes an graph-based clustering approach compared to (Macosko et al.). Importantly, the distance metric which drives the clustering analysis (based on previously identified PCs) remains the same. However, our approach to partioning the cellular distance matrix into clusters has dramatically improved. Our approach was heavily inspired by recent manuscripts which applied graph-based clustering approaches to scRNA-seq data [SNN-Cliq, Xu and Su, Bioinformatics, 2015](https://academic.oup.com/bioinformatics/article/31/12/1974/214505) and CyTOF data [PhenoGraph, Levine et al., Cell, 2015](http://www.ncbi.nlm.nih.gov/pubmed/26095251). Briefly, these methods embed cells in a graph structure - for example a K-nearest neighbor (KNN) graph, with edges drawn between cells with similar gene expression patterns, and then attempt to partition this graph into highly interconnected ‘quasi-cliques’ or ‘communities’. As in PhenoGraph, we first construct a KNN graph based on the euclidean distance in PCA space, and refine the edge weights between any two cells based on the shared overlap in their local neighborhoods (Jaccard distance). To cluster the cells, we apply modularity optimization techniques [SLM, Blondel et al., Journal of Statistical Mechanics](http://dx.doi.org/10.1088/1742-5468/2008/10/P10008), to iteratively group cells together, with the goal of optimizing the standard modularity function.

The <code>FindClusters</code> function implements the procedure, and contains a resolution parameter that sets the ‘granularity’ of the downstream clustering, with increased values leading to a greater number of clusters. We find that setting this parameter between 0.6-1.2 typically returns good results for single cell datasets of around 3K cells. Optimal resolution often increases for larger datasets. The clusters are saved in the <code>object@ident</code> slot.

```{r SNN}
# save.SNN = T saves the SNN so that the clustering algorithm can be rerun
# using the same graph but with a different resolution value (see docs for
# full details)

obj <- FindClusters(object = obj, 
                    reduction.type = "pca", 
                    dims.use = 1:10, 
                    resolution = 0.6, 
                    print.output = 0, 
                    save.SNN = TRUE)

```

A useful feature in Seurat v2.0 is the ability to recall the parameters that were used in the latest function calls for commonly used functions. For FindClusters, we provide the function PrintFindClustersParams to print a nicely formatted formatted summary of the parameters that were chosen.

```{r PrintClusters}
PrintFindClustersParams(object = obj)
```


## Run Non-linear dimensional reduction (tSNE)

Seurat continues to use tSNE as a powerful tool to visualize and explore these datasets. While we no longer advise clustering directly on tSNE components, cells within the graph-based clusters determined above should co-localize on the tSNE plot. This is because the tSNE aims to place cells with similar local neighborhoods in high-dimensional space together in low-dimensional space. As input to the tSNE, we suggest using the same PCs as input to the clustering analysis, although computing the tSNE based on scaled gene expression is also supported using the genes.use argument.

```{r RunTSNE}
obj <- RunTSNE(object = obj, dims.use = 1:10, do.fast = TRUE)
# note that you can set do.label=T to help label individual clusters
TSNEPlot(object = obj)
```

## Finding differentially expressed genes (cluster biomarkers)

Seurat can help you find markers that define clusters via differential expression. By default, it identifes positive and negative markers of a single cluster (specified in <code>ident.1</code>), compared to all other cells. <code>FindAllMarkers</code> automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.

The min.pct argument requires a gene to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a gene to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of genes that are unlikely to be highly discriminatory. As another option to speed up these computations, max.cells.per.ident can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significiant and the most highly differentially expressed genes will likely still rise to the top.

```{r Cluster1Markers}
# find all markers of cluster 1
cluster1.markers <- FindMarkers(object = obj, 
                                ident.1 = 1, 
                                min.pct = 0.25)
print(x = head(x = cluster1.markers, n = 5))
```

```{r Cluster5Markers}
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers <- FindMarkers(object = obj, 
                                ident.1 = 5, 
                                ident.2 = c(0, 3), 
                                min.pct = 0.25)
print(x = head(x = cluster5.markers, n = 5))
```

```{r FindMarkers}
# find markers for every cluster compared to all remaining cells, report
# only the positive ones
obj.markers <- FindAllMarkers(object = obj, 
                               only.pos = TRUE, 
                               min.pct = 0.25, 
                               thresh.use = 0.25)

obj.markers %>% 
  group_by(cluster) %>% 
  top_n(2, avg_logFC)

gene_list <- lapply(unique(obj.markers$cluster), function(c){

obj.markers %>% 
  group_by(cluster) %>% 
  top_n(2, avg_logFC) %>%
  filter(cluster == c) %>% .$gene

})

```

Seurat has four tests for differential expression which can be set with the test.use parameter: ROC test (“roc”), t-test (“t”), LRT test based on zero-inflated data (“bimod”, default), LRT test based on tobit-censoring models (“tobit”) The ROC test returns the ‘classification power’ for any individual marker (ranging from 0 - random, to 1 - perfect).

```{r, eval = FALSE}
cluster1.markers <- FindMarkers(object = obj, 
                                ident.1 = 0, 
                                thresh.use = 0.25, 
                                test.use = "roc", 
                                only.pos = TRUE)
```

We include several tools for visualizing marker expression. <code>VlnPlot</code> (shows expression probability distributions across clusters), and <code>FeaturePlot</code> (visualizes gene expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring <code>JoyPlot</code>, <code>CellPlot</code>, and <code>DotPlot</code> as additional methods to view your dataset.

### Violin Plot (Cluster 0)
```{r ViolinPlot}
VlnPlot(object = obj, features.plot = gene_list[[1]])
# for( i in seq_along(lst) ) VlnPlot(object = obj, features.plot = gene_lst[[i]])
```
### Violin Plot Raw Count (Cluster 0)
```{r ViolinRaw}
# you can plot raw UMI counts as well
VlnPlot(object = obj, 
        features.plot = gene_list[[1]], 
        use.raw = TRUE, 
        y.log = TRUE)
```

### Top Gene in Each Cluster
```{r TopGene}
FeaturePlot(object = obj, 
            features.plot = sapply(seq_along(gene_list), function(i) gene_list[[i]][1]), 
            cols.use = c("grey", "blue"), 
            reduction.use = "tsne")
```

### Top 10 Genes by Cluster Heatmap
<code>DoHeatmap</code> generates an expression heatmap for given cells and genes. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.

```{r ClusterHeatmap}
top10 <- obj.markers %>% 
  group_by(cluster) %>% 
  top_n(10, avg_logFC)

# setting slim.col.label to TRUE will print just the cluster IDS instead of
# every cell name
DoHeatmap(object = obj, 
          genes.use = top10$gene, 
          slim.col.label = TRUE, 
          remove.key = TRUE)
```

## Assigning cell type identity to clusters

This code is part of the Seurat tutorial and is left in for reference.  It is not run because these gene markers may not be available for you data set.  

Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:

```{r, eval = FALSE}
current.cluster.ids <- c(0, 1, 2, 3, 4, 5, 6, 7)
new.cluster.ids <- c("CD4 T cells", "CD14+ Monocytes", "B cells", "CD8 T cells", 
    "FCGR3A+ Monocytes", "NK cells", "Dendritic cells", "Megakaryocytes")
obj@ident <- plyr::mapvalues(x = obj@ident, 
                             from = current.cluster.ids, 
                             to = new.cluster.ids)
TSNEPlot(object = obj, 
         do.label = TRUE, 
         pt.size = 0.5)
```


## Further subdivisions within cell types

If you perturb some of our parameter choices above (for example, setting <code>resolution=0.8</code> or changing the number of PCs), you might see the CD4 T cells subdivide into two groups. You can explore this subdivision to find markers separating the two T cell subsets. However, before reclustering (which will overwrite <code>object@ident</code>), we can stash our renamed identities to be easily recovered later.

```{r, eval = FALSE}
# First lets stash our identities for later
obj <- StashIdent(object = obj, save.name = "ClusterNames_0.6")

# Note that if you set save.snn=T above, you don't need to recalculate the
# SNN, and can simply put: pbmc <- FindClusters(pbmc,resolution = 0.8)
obj <- FindClusters(object = obj, 
                    reduction.type = "pca", 
                    dims.use = 1:10, 
                    resolution = 0.8, 
                    print.output = FALSE)
```

```{r, eval = FALSE}
# Demonstration of how to plot two tSNE plots side by side, and how to color
# points based on different criteria
plot1 <- TSNEPlot(object = obj, 
                  do.return = TRUE, 
                  no.legend = TRUE, 
                  do.label = TRUE)
plot2 <- TSNEPlot(object = obj, 
                  do.return = TRUE, 
                  group.by = "ClusterNames_0.6", 
                  no.legend = TRUE, 
                  do.label = TRUE)
plot_grid(plot1, plot2)
```

```{r, eval = FALSE}
# Find discriminating markers
tcell.markers <- FindMarkers(object = obj, 
                             ident.1 = 0, 
                             ident.2 = 1)

# Most of the markers tend to be expressed in C1 (i.e. S100A4). However, we
# can see that CCR7 is upregulated in C0, strongly indicating that we can
# differentiate memory from naive CD4 cells.  cols.use demarcates the color
# palette from low to high expression
FeaturePlot(object = obj, 
            features.plot = c("S100A4", "CCR7"), 
            cols.use = c("green", "blue"))
```

The memory/naive split is bit weak, and we would probably benefit from looking at more cells to see if this becomes more convincing. In the meantime, we can restore our old cluster identities for downstream processing.

```{r, eval = FALSE}
obj <- SetAllIdent(object = obj, id = "ClusterNames_0.6")
```

```{r SaveObject}
obj@misc <- params$path
out_file <- paste0("data/", paste0(params$project, collapse = "_"), ".rds")
saveRDS(obj, file = out_file)
```

